### PR TITLE
Add unified i/o and file tracking with SourceMap

### PR DESCRIPTION
**Pull request by Dari-OS**
## Summary
This PR introduces the foundational `source` crate, which is responsible for managing source code, handling file I/O, and providing a unified way to track source locations (`Span`s) across the compiler.

This infrastructure decouples file loading/reading from the core compiler logic, enabling a central and robust reporting for [error handling](https://github.com/Randware/Wasome/tree/feature/error-handling) purposes.

### 1. Core Source Management (`SourceMap` & `SourceFile`)
* Implemented `SourceMap` as the central registry for loaded files, handling deduplication and caching.
* Implemented `SourceFile` with pre-computed line indices.
* Added support for **Unicode** column offsets (handling multi-byte characters like ✨ correctly).
* Added **Newline Normalization** (Windows `\r\n` -> `\n`) to ensure consistent offsets. 

### 2. File Loading Abstraction (`FileLoader`)
* Defined the `FileLoader` trait to abstract file system operations. (Makes it theoretically possible to load files from an `URL` )
* Implemented `WasomeLoader` for real disk I/O with `fs::canonicalize` and safety checks.
* Enforced a **4GB hard limit** on single file sizes to optimize memory usage (see Design Choices).

### 3. Types 
* Introduced `BytePos` and `FileID` as `u32` wrappers.
* Implemented `Span` with **exclusive ranges** (`start..end`).
* Added helpers for `Span` merging, containment checks, and source slicing.

## Design Choices
* **`u32` for Positions:** We use `u32` instead of `usize` for `BytePos` and `Span`. This reduces the size of a `Span` from 24 bytes (on 64-bit) to **12 bytes**. This creates a hard limit of 4GB per source file, which is an acceptable trade-off for the massive memory savings.
* **Trait-based Loading:** By using a generic `Loader`, we can swap the real filesystem for a `MockLoader` in tests, allowing us to test edge cases (like missing files or permissions) without touching the actual disk.

## Testing Strategy
* **Unit Tests:** comprehensive coverage of arithmetic, span logic, and line calculation math.
* **Integration Tests:** End-to-end verification using `tempfile` to simulate real `.waso` files on disk.
* **Mock Tests:** Verified caching and virtual file loading using an in-memory `HashMap` via `MockLoader`.

## Checklist
- [x] All unit and integration tests pass.
- [x] Documentation added for public (and internal) APIs.

### CODE REVIEW

File: compiler/source/src/loader.rs [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    #[test]
+    fn test_resolve_happy_path() {
+        let dir = tempdir().unwrap();
+        let root = dir.path();
+
+        // Create a real file so canonicalize works
+        // NOTE: A MockFS would not work because the real
+        // Wasome loader canonicalizes the paths, which means all files
+        // have to exist
+        let file_path = root.join("main.wasom");
```
Comment (Dari-OS) [inline compiler/source/src/loader.rs]:
```suggestion
        let file_path = root.join("main.waso");
```

File: compiler/source/tests/integration_tests.rs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    write!(file, "{}", content).unwrap();
+    (dir, file_path)
+}
+
+// --- INTEGRATION TESTS ---
+
+#[test]
+fn test_workflow_load_and_lookup() {
+    let content = "fn main() {\n    let x = 10;\n}";
+    let (dir, _path) = setup_file("main.waso", content);
```
[the remark this answers is not in the document]
Comment (Dari-OS) [reply compiler/source/tests/integration_tests.rs]: This is wrong, `dir` gets used right underneath this statement

File: compiler/source/src/source.rs [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+        // Access to accumulated gap of the nerest mb
+        let gap = if idx > 0 {
+            line_info.multi_byte_chars[idx - 1].accumulated_gap
+        } else {
+            0
+        };
+
+        // Safety: When the provided span would have a start INSIDE a multi byte char
+        // this - would panic if we don't use saturating_sub because the result would be negative
+        (line_index as u32 + 1, byte_col.saturating_sub(gap) + 1)
```
Comment (Running-Lag) [inline compiler/source/src/source.rs]: This seems like it doesn't work properly when pos is inside a `MultiByteChar` with `byte_len > 2` and not at the start as we correct for the entire length despite parts of the length being after the provided `pos`.

For example, we have only one line with a single byte char and then a char with a width of 4 bytes and provide a pos of 2 (second byte of the 4-byte char).
We would then do `saturating_sub(2, 3 /* 4 - 1 */)`, which results in 0 instead of the expected 1.

File: compiler/source/src/source.rs [hunk 4]
```diff
@@ -0,0 +1,558 @@
...
+
+    /// Takes a [pos](BytePos) and a [file](SourceFile) and coverts it to a [location](Location).
+    ///
+    /// # Returns
+    ///
+    /// * `Location`- The 1-based [`Location`]
+    ///
+    /// # Warning
+    ///
+    /// If the [`BytePos`] does not belong to [file](SourceFile) **the returned data
```
[the remark this answers is not in the document]
Comment (Dari-OS) [reply compiler/source/src/source.rs]: This is wrong.
(see: https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html)

File: compiler/source/src/source.rs (hunk 3 above)
Comment (Dari-OS) [reply compiler/source/src/source.rs]: Woah, nice observation! 
Fixed.

File: compiler/source/src/source.rs [hunk 5]
```diff
@@ -0,0 +1,558 @@
...
+
+    /// Loads the file to the [`SourceMap`]
+    ///
+    /// Takes a relative path to a file, adds it to the [cache](SourceMap::file_cache),
+    /// loads that content of the file into a [`SourceFile`] and adds the [file](SourceFile) to the [SourceMap](SourceMap::files)
+    ///
+    /// # Returns
+    ///
+    /// * `Ok(FileID)`- A handle to the newly *stored* file
+    /// * `Err()` - An i/o [error](std::io::Error) in case the underlying `OS` returns an [error](std::io::Error)
```
[the remark this answers is not in the document]
Comment (Dari-OS) [reply compiler/source/src/source.rs]: Fixed.

File: compiler/source/src/types.rs [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+            return None;
+        }
+
+        Some(Span {
+            file_id: self.file_id,
+            start: cmp::min(self.start, other.start),
+            end: cmp::max(self.end, other.end),
+        })
+    }
+}
```
[the remark this answers is not in the document]
Comment (Dari-OS) [reply compiler/source/src/types.rs]: > The best way to solve this would probably be to forbid the merging of such spans entirely as it leads to more-than-unintuitive behavior.

Fair point on the naming. Calling it a "union" is definitely confusing since we're implicitly filling in the gaps.

However, for the AST, We actually *need* this specific behavior. Think about parsing `left + right`. We need a span that covers from the start of `left` to the end of `right`, including the whitespace/operator in between. If we forbid merging disjoint spans, we can't create parent nodes for things like binary expressions.

I'm going to keep the logic (it's pretty standard in compiler tooling), but I'll update the docs to clarify it calculates the **convex hull** or **cover** rather than a strict set union. Thanks!

[the remark this answers is not in the document]
Comment (Running-Lag) [reply compiler/source/src/types.rs]: Fair point
